### PR TITLE
Build `rust_test` targets with `crate` using the same crate name as the underlying library target.

### PR DESCRIPTION
**Pull request by cfredric**
This PR changes `rust_test` to use the same crate name as the underlying target, in cases where the `crate` attribute is used. This is consistent with `cargo`, which builds the crate using the name of the library when running `cargo test`.

The behavior of `rust_test` targets that don't use the `crate` attribute is unchanged.

### CODE REVIEW

**Comment by cfredric** [comment]
Looks like there's an issue with compile_data, but I don't immediately see why. I'd have expected things to work fine, unless Bazel is overwriting the output of one target with the output of another, due to the test using the same crate name as the library?

**Comment by cfredric** [comment]
FWIW, I believe the failure to link on some windows CI bots is some transient issue - it's not a consistent failure.

**Comment by cfredric** [comment]
Friendly ping, could someone please take a look at this?

**Comment by UebelAndre** [comment]
@scentini were you gonna take a look?

File: test/rust/src/lib.rs [hunk 1]
```diff
@@ -12,4 +12,5 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#![crate_name = "hello_lib"]
```
Comment (scentini) [inline test/rust/src/lib.rs:15]: Why do we need this change?

**Review by scentini — COMMENTED** [review]
Except for the one question I have, LGTM.

**Comment by scentini** [comment]
Sorry for the delay, I thought I'd already reviewed this PR.

**Pull request merged**